syntax = "proto3";



enum MessageType {
    MSG_NONE            = 0;

    MSG_PING_REQUEST    = 1;
    MSG_PING_RESPONSE   = 2;
    
    MSG_BMS_REQUEST     = 3;
    MSG_BMS_RESPONSE    = 4;
    
    MSG_IMU_REQUEST     = 5;
    MSG_IMU_RESPONSE    = 6;
    
    MSG_MOTOR_REQUEST   = 7;
    MSG_MOTOR_RESPONSE  = 8;
}

message EmptyResponse {
   MessageType type    = 1;
}

// ping test
message PingRequest {
    MessageType type    = 1;
    int32 data_0        = 2;
    int32 data_1        = 3;
    int32 data_2        = 4;
    int32 data_3        = 5;
}

message PingResponse {
    MessageType type    = 1;
    int32 data_0        = 2;
    int32 data_1        = 3;
    int32 data_2        = 4;
    int32 data_3        = 5;
}


// baterry power management
message BMSRequest {
    MessageType type        = 1;

    bool main_bus_a_enable  = 2;
    bool main_bus_b_enable  = 3;
}

message BMSResponse { 
    MessageType type            = 1;  

    int32 cell_voltage_1        = 2;
    int32 cell_voltage_2        = 3;
    int32 cell_voltage_3        = 4;   

    int32 main_bus_a_voltage    = 5;
    int32 main_bus_a_current    = 6;

    int32 main_bus_b_voltage    = 7;
    int32 main_bus_b_current    = 8;

    bool main_bus_a_enabled     = 9;
    bool main_bus_b_enabled     = 10;
}

// IMU + GPS unit
message IMURequest {
    MessageType type    = 1;
}

message IMUResponse {
    MessageType type    = 1;
   
    int32 roll          = 2; 
    int32 pitch         = 3; 
    int32 yaw           = 4;

    int32 gps_x         = 5;
    int32 gps_y         = 6;
    int32 gps_h         = 7;
}

// motor controll unit
message MotorRequest {
    MessageType type            = 1;

    int32 required_velocity     = 2;
    int32 maximum_velocity      = 3;
    int32 maximum_acceleration  = 4;
}

message MotorResponse {
    MessageType type            = 1;

    int32 current_velocity      = 2; 
    int32 current_encoder       = 3; 
}
